serialise/src/Codec/Serialise/Class.hs

```diff
@@ -332,22 +333,16 @@ instance Serialise Char where
 
     -- For [Char]/String we have a special encoding
     encodeList cs
-      | any isSurrogate cs = do
-            encodeListLen (fromIntegral (length cs) + 1)
-            <> encodeWord 0
-            <> foldMap (encodeWord . fromIntegral . ord) cs
+      | any isSurrogate cs = encodeByteArray (encodeGenUTF8 cs)
```

Comment:
So ideally I'd like us to be able to do this:

```
encodeList cs =
    case encodeGeneralisedUTF8 cs of
      (ba, len, off, isUTF8Subset)
        |  isUTF8Subset -> encodeUTF8ByteArray ba len off
        | otherwise     -> encodeByteArray ba len off
```
since that lets us do it in one pass, and when it is in the UTF8 subset we can just reuse the other existing primitives.